crates/sparrow-syntax/src/syntax/signature.rs

```diff
@@ -1,7 +1,53 @@
+use hashbrown::HashSet;
 use itertools::Itertools;
 
 use crate::parser::try_parse_signature;
-use crate::{ExprRef, FeatureSetPart, FenlType, Located, Parameters, ParseErrors};
+use crate::{ExprRef, FeatureSetPart, FenlType, Located, Parameters, ParseErrors, TypeConstraint};
+
+/// Type variable defined by a signature
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+#[cfg_attr(test, derive(serde::Serialize))]
+pub struct TypeVariable(pub String);
+
+impl std::fmt::Display for TypeVariable {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+/// Contains the type variable and the constraints for
+/// that variable.
+///
+/// e.g. N(type variable): Eq(constraint) + Hash(constraint)
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+#[cfg_attr(test, derive(serde::Serialize))]
+pub struct TypeParameter {
+    pub name: TypeVariable,
+    pub constraints: Vec<TypeConstraint>,
+}
+
+impl std::fmt::Display for TypeParameter {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // Note: this excludes the type constraints
+        write!(f, "{}", self.name.0)
+    }
+}
+
+impl TypeParameter {
+    pub fn new(name: String, constraints: Vec<TypeConstraint>) -> Self {
+        Self {
+            name: TypeVariable(name),
+            constraints,
+        }
+    }
+
+    pub fn try_new(name: &str, constraint: &TypeConstraint) -> anyhow::Result<Self> {
```

Comment:
What's the purpose of this? When would it ever fail?

Reply:
This is used in `grammar.lalrpop`. It used to be able to fail when I was passing `FenlType` but now that I explicitly use `TypeConstraint` I suppose I can just remove this.